### ISSUE TITLE
More effective backsubstitution in solve_poly_system()

### ISSUE THREAD

**Issue by filip-misun** [sympy/sympy#14805]
The solve_poly_system function, which is used for solving systems of polynomial equations, currently implements the backsubstitution method: construct groebner basis w.r.t. the lexicographic ordering and then use backsubstitution to find all solutions, starting with the univariate basis polynomial. If you look into the implementation of solve_poly_system, you can see that at each step of the backsubstitution, the groebner basis is computed again from scratch. Computation of groebner basis can be very expensive and you don't want to do it more than once if it's not necessary. In this case, it is quite easy to implement the backsubstitution method in a way that constructs the groebner basis only once.

So unless there is a reason why the groebner basis is computed from scratch at each step, I suggest I could fix this.

**Comment by avishrivastava11** [comment]
i would like to take up this issue if open..

**Comment by avishrivastava11** [comment]
@filip-misun  can you tell the path (from the main repo ==> into the exact file) of the file and the exact location of the code (snippet) where you are suggesting the improvement

**Comment by skirpichev** [comment]
Oh...

@filip-misun, AFAIK it's not a requirement that you must fill bugreport first.  If you have a clear idea what to do - just make a pull request.  Beware, backsubstitution algorithm of solve_poly_system() in sympy is not very well developed (i.e. it can handle only zero-dimensional systems, can omit some solutions due to using roots() to solve univariate polynomial equations, etc).  So, you may want to look on this function in the [diofant](https://github.com/diofant/diofant).

**Comment by filip-misun** [comment]
> i would like to take up this issue if open..

Sorry @avishrivastava11, I am already working on it and it's almost done.

> AFAIK it's not a requirement that you must fill bugreport first. If you have a clear idea what to do - just make a pull request.

I am just following the instructions in the [Development workflow](https://github.com/sympy/sympy/wiki/Development-workflow#pick-an-issue-to-fix), where it says "Create an issue if there isn't one for what you want to do". I guess I could skip this step, but honestly, I don't think it is a bad practice.

> Beware, backsubstitution algorithm of solve_poly_system() in sympy is not very well developed (i.e. it can handle only zero-dimensional systems, can omit some solutions due to using roots() to solve univariate polynomial equations, etc).

Yeah, I can see that. The function roots() will indeed omit solutions which cannot be expressed with radicals. However, I guess this problem should be fixed in the roots() function (if it can be fixed somehow). I also believe that the user should have the option to have the system of polynomial equations solved numerically. This will give only approximate solutions, but won't miss any of them. There is the nsolve() function, but AFAIK, this can handle only one univariate polynomial. But for numerical approximation, I believe that the eigenvalue method is more appropriate than the backsubstitution method. I have already opened an issue on this topic #14799.

It would be awesome if solve_poly_system() could also handle system that are not zero-dimensional. But how do you propose this could be solved? Do you mean some kind of parametrization of the system? As far as I know, this is a really difficult problem, and isn't quite solved to this day. Does diofant solve this somehow? We could open another issue on this topic, if you have any ideas.

**Comment by asmeurer** [comment]
There's no problem with opening an issue first. But if you already have some work in progress code you can start a WIP pull request and we can discuss things there (it's usually easier to discuss ideas if there is some code, even if it's incomplete). 

By the way, `nsolve` can handle systems of equations just fine (it uses newton's method, via mpmath.findroot).

**Comment by skirpichev** [comment]
> here it says "Create an issue if there isn't one for what you want to do".

I see, sorry for misinformation.  That's something new, I guess to reduce numbers of stale PRs in the sympys review queue...

> I guess this problem should be fixed in the roots() function

I'm not sure.  In the diofant this is fixed by using .all_roots() method, instead of roots().

> This will give only approximate solutions, but won't miss any of them. There is the nsolve() function

Oh, I bet - nsolve() can miss solutions.  (but I think, that mpmath.polyroots shouldn't)

> But how do you propose this could be solved?

You can take look on the code, if interested.  Method is same as implemented e.g. in Reduce (Herbert Melenk, "Algebraic Solution of Nonlinear Equation Systems in REDUCE").

**Comment by filip-misun** [comment]
Sorry, I didn't express myself quite clearly. Of course, there is no numerical algorithm that guarantees to find all roots of any instance of a polynomial. But there are algorithms that perform reasonably well in finding all roots, except for some pathological instances. Numpy, for example, has polyroots() function, which reduces the root location problem to an eigenvalue problem of the companion matrix. You could also find all real roots using Sturm sequences and bisection. But I don't really know what the state-of-the-art is in this topic, I am just a computer science college student :)
